ynr/apps/sopn_parsing/helpers/pdf_helpers.py

```diff
@@ -19,20 +21,106 @@
 
 
 class SOPNDocument:
-    def __init__(self, file):
+    def __init__(self, file, all_documents_with_source):
+        """
+        Represents a collection of pages from a single PDF file.
+        """
         self.file = file
-        self.pages = []
-        self.parse_pages()
+        self.unmatched_documents = set(doc for doc in all_documents_with_source)
+        self.matched_documents = set()
+        self.pages = self.parse_pages()
         self.document_heading = self.pages[0].get_page_heading_set()
+
         if len(self.document_heading) < 10:
             raise NoTextInDocumentError()
-        self.top_pages = [
-            p.page_number
-            for p in self.pages
-            if p.detect_top_page(self.document_heading)
+
+    @property
+    def matched_pages(self):
+        return [p for p in self.pages if p.matched]
+
+    @property
+    def unmatched_pages(self):
+        return [p for p in self.pages if not p.matched]
+
+    def validate_page_numbers(self, page_numbers):
+        if len(page_numbers) == 1:
+            return True
+        page_numbers = [page for page in page_numbers]
+        lower = min(page_numbers)
+        upper = max(page_numbers) + 1
+
+        assert page_numbers == list(range(lower, upper))
+
+    def match_ballot_to_pages(self, ballot: Ballot) -> str:
+        """
+        Given a ballot object, loop over all the unmatched pages looking for the
+        first match.
+
+        Return a string in the format of:
+           "1,2,3,4"
+        OR
+            "all"
+
+        For more on this format, see https://camelot-py.readthedocs.io/en/master/user/quickstart.html?highlight=pages#specify-page-numbers
+        """
+        page_numbers = []
+        previous_page = None
+        for page in self.unmatched_pages:
+            post_label = ballot.post.label
+            if previous_page:
+                if page.is_continuation_page(
+                    self.document_heading, previous_page, post_label
+                ):
+                    page_numbers.append(page.page_number)
+                else:
+                    break
+            else:
+                if page.matches_ward_name(post_label):
+                    page_numbers.append(page.page_number)
+                    page.matched = True
+                    previous_page = page
+        if self.validate_page_numbers(page_numbers):
+            return ",".join(str(p) for p in page_numbers)
+        else:
+            raise ValueError("Page numbers are not consecutive")
+
+    def match_all_pages(self) -> List[Tuple[OfficialDocument, str]]:
+        """
+
+        [
+            (doc, "123"),
+            (doc, "456"),
         ]
 
+        :return:
+        """
+        if len(self.pages) == 1 and len(self.unmatched_documents) == 1:
+            return [(self.unmatched_documents.pop(), "all")]
+
+        docs_by_sorted_ballot_label = sorted(
```

Comment:
Not sure this is needed as ordering is already applied when we get the `OfficialDocument` objects earlier on https://github.com/DemocracyClub/yournextrepresentative/pull/1659/files#

Reply:
Yep, I just wanted to make really sure they were sorted. It feels better to do this "inside the black box" rather than trusting it to be sorted in advance.

In fact, I wonder if taking a file and a source would be better for this class, and then we can do the query in-class? What do you think?

Reply:
Yeah I think that's a good idea - that would allow us to simplify the logic  in `extract_pages_for_single_document` also